provider/file/file.go

```diff
@@ -18,68 +22,135 @@ var _ provider.Provider = (*Provider)(nil)
 // Provider holds configurations of the provider.
 type Provider struct {
 	provider.BaseProvider `mapstructure:",squash"`
+	Directory             string `description:"Load configuration from one or more .toml files in a directory"`
 }
 
 // Provide allows the file provider to provide configurations to traefik
 // using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
-	watcher, err := fsnotify.NewWatcher()
+	var configuration *types.Configuration
+	var err error
+	var watchDir string
+
+	if p.Directory != "" {
+		watchDir = p.Directory
+		configuration, err = loadFileConfigFromDirectory(p.Directory)
+	} else {
+		watchDir = filepath.Dir(p.Filename)
+		configuration, err = loadFileConfig(p.Filename)
+	}
+
 	if err != nil {
-		log.Error("Error creating file watcher", err)
 		return err
 	}
 
-	file, err := os.Open(p.Filename)
-	if err != nil {
-		log.Error("Error opening file", err)
+	if err := p.addWatcher(pool, watchDir, configurationChan, p.watcherCallback); err != nil {
 		return err
 	}
-	defer file.Close()
-
-	if p.Watch {
-		// Process events
-		pool.Go(func(stop chan bool) {
-			defer watcher.Close()
-			for {
-				select {
-				case <-stop:
-					return
-				case event := <-watcher.Events:
-					if strings.Contains(event.Name, file.Name()) {
-						log.Debug("Provider event:", event)
-						configuration := p.loadFileConfig(file.Name())
-						if configuration != nil {
-							configurationChan <- types.ConfigMessage{
-								ProviderName:  "file",
-								Configuration: configuration,
-							}
-						}
-					}
-				case error := <-watcher.Errors:
-					log.Error("Watcher event error", error)
+
+	sendConfigToChannel(configurationChan, configuration)
+	return nil
+}
+
+func (p *Provider) addWatcher(pool *safe.Pool, directory string, configurationChan chan<- types.ConfigMessage, callback func(chan<- types.ConfigMessage, fsnotify.Event)) error {
+	// Debounce used to ensure that multiple watcher events in a short space of time don't trigger multiple config reloads
+	var debouncePeriod = 1 * time.Second
+	var event *fsnotify.Event
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating file watcher: %s", err)
+	}
+
+	// Process events
+	pool.Go(func(stop chan bool) {
+		defer watcher.Close()
+		for {
+			select {
+			case <-stop:
+				return
+			case evt := <-watcher.Events:
+				event = &evt
+			case <-time.After(debouncePeriod):
```

Comment:
@rjshep I'm not sure it's really needed knowing that indeed we have a similar global mechanism in [server.go](https://github.com/containous/traefik/blob/master/server/server.go#L238). I think we should simplify this kind of filtering and keep it one place only :)